refactor: Fix TokenRateLimitPolicy tests failing in parallel execution

Anton-Fil opened the pull request:
## Description
closes #769 

**Problem Description**
The TokenRateLimitPolicy tests in both `basic/ and multiple_iterations/ directories `were failing when executed in parallel, while working correctly when run sequentially. The issue occurred because:

1. The basic tests `(test_trlp.py, test_trlp_streaming.py)` use shared authentication fixtures with identical user IDs, while multiple iterations tests use anonymous requests but still compete for the same rate limit quota
2. The TokenRateLimitPolicy is configured to track rate limits per user ID using `counters=[CelExpression("auth.identity.userid")]` for basic tests, and per IP/request for multiple iterations tests
3. When tests run in parallel, they simultaneously consume tokens from shared limits (50 tokens/30s for basic tests, 50 tokens/20s for multiple iterations), causing `unexpected 200 `responses instead of the `expected 429` rate limit responses, as the tests were interfering with each other's token consumption patterns

**Attempted Solutions (That Didn't Work)**
Several approaches were considered but not implemented:

- Attempted to use `pytest-xdist` worker IDs to create unique user IDs per worker, but this approach failed
- Tried creating separate user fixtures for each test file, but this required changes to the shared conftest.py and authorization configuration that weren't feasible
- Attempted to set up different backends for each test module to isolate rate limiting, but this approach gave the same failures and added complexity to the setup

**Final Solution**
Modified the test logic to be more resilient to parallel execution by:

- Changed the while loops to bounded for loops so the test now stops when a `429` response is received, instead of waiting to reach the exact token limit. The original while `total_tokens < LIMIT` loop relied on precise token accounting, but in parallel runs the shared rate-limit counters caused small timing differences between workers. This led to inconsistent token tracking, where some threads consumed tokens faster or slower, causing the loop to exit early before the actual limit was reached.
- Set reasonable upper bounds (`max_requests = 20` for most tests, `max_requests = 30` for paid user tests) to prevent infinite loops
- Tests now break the loop immediately when rate limit (429) is encountered, rather than continuing to consume tokens
- Added `pytest.fail()` with descriptive messages when rate limits aren't hit within the expected request count
- Added a readiness check to make sure the route is ready before committing the policy, preventing “route not attached to existing path” errors.
- Reduced the iteration range (from 10→5 and 5→3) to speed up test execution while still running enough cycles to verify TokenRateLimitPolicy reset and enforcement functionality

emmaaroche commented on testsuite/tests/singlecluster/limitador/tokens_rate_limit/multiple_iterations/test_multiple_trlp_iterations.py:
```diff
@@ -20,28 +19,39 @@
         assert (
             response.status_code == 429
-        ), f"Iteration {i+1}/10: Expected 429 after {total_tokens}/{LIMIT.limit} tokens, but got {response.status_code}"
+        ), f"Iteration {i+1}/5: Expected 429 to still be enforced, but got {response.status_code}"
```
``` suggestion
        ), f"Iteration {i+1}/5: Expected 429 after {total_tokens}/{LIMIT.limit} tokens, but got {response.status_code}"

Anton-Fil commented:
Hey @emmaaroche , I applied the suggested changes but the test fails with:

`Failed: Never hit rate limit after 15 requests.`

I also refined tests further for better stability. Here’s what I added:

- `module_label` to user IDs to isolate rate limit counters per worker — this should prevent counter conflicts in parallel runs.
- also i added a` hit_limit` flag to explicitly verify that 429 responses were received, ensuring more deterministic checks for rate limit enforcement.

azgabur reviewed (commented):
I think the whole problem comes down to the `token_rate_limit` fixture and the structure of test modules. The fixture is `module` scoped. This can be problematic for rate limit tests as multiple test functions inside one module will use the same TRLP which obviously will cause interference. Thats why most limitador tests contain only one test function per module. 
To keep the current setup I suggest changing the scope for `token_rate_limit` and `commit` fixtures to `function`. That will force re-creation of TRLP and AuthPolicy for each test function effectively re-setting the counter but at the cost of re-creating and waiting for reconcile.

Another solution is to separate test functions to multiple modules, that will cause the xdist library to run these in parallel as well as each having their own TRLP which may be in the end faster as it divides the work among workers per-module basis. I will leave that decision on you. But especially if you use sleeps than parallelization should be preferred. 

See how the changed scope for RLP and commit fixture was done in: `testsuite/tests/singlecluster/limitador/storage`

I tested this simple change on `main` but I still got some test fails so its probably not this simple.

emmaaroche commented:
Thanks for the suggestions @azgabur. Following your feedback, @Anton-Fil and I experimented with several approaches:

- Created separate test modules for each test case
- Defined a unique TRLP per module
- Used different limit configurations per TRLP
- Tried switching to function-scoped fixtures
- Tested separate TRLPs for route- and gateway-level setups

Unfortunately, the same TRLP issues are still showing up after making these changes 😢

crstrn13 commented:
@emmaaroche @Anton-Fil 
I have noticed that in the `multiple-iterations/conftest.py` we do not set any counter on the TRLP. Is this by design or is it interfering with the test logic itself if we added a counter there? The counter doesn't have to be auth-specific, I think we could add a header to all of the requests in that test for instance and count on that, for instance.

Thing is I liked more the way we asserted for roughly the number of tokens, rather that sending requests until we run out of quota. Does that make any sense? Let me know

emmaaroche commented on testsuite/tests/singlecluster/limitador/tokens_rate_limit/basic/test_trlp.py:
```diff
@@ -48,22 +48,30 @@ def test_trlp_limit_and_reset_free_user(client, free_user_auth):
         assert tokens_used > 0 and tokens_used == usage["prompt_tokens"] + usage["completion_tokens"]
         total_tokens += tokens_used
+    else:
+        pytest.fail(f"Never hit rate limit after {max_requests} requests")
```
I think you could remove the `else` after the request loop in each test as the `assert hit_limit` already covers the failure case if the rate limit isn’t reached.
